client/network/src/behaviour.rs

```diff
@@ -148,6 +150,18 @@ pub enum BehaviourOut<B: BlockT> {
 	/// Events generated by a DHT as a response to get_value or put_value requests as well as the
 	/// request duration.
 	Dht(DhtEvent, Duration),
+
+	/// Event generated by bitswap.
+	Bitswap(BitswapEvent)
+}
+
+/// An event generated by bitswap.
+#[derive(Clone, Debug)]
+pub enum BitswapEvent {
+	/// A block was received.
+	ReceivedBlock(PeerId, cid::Cid, Box<[u8]>),
```

Comment:
what's the difference between a `Box<[u8]>` and a `Vec<u8>`

Reply:
The main thing is that it's the type that `libp2p_bitswap` uses. Another thing is that you know that all the allocated bytes are used, as opposed to a `Vec` where the allocated `capacity` could be a lot bigger than the `len`.

Reply:
I don't think it's true, when converting a Vec to a Box the capacity isn't truncated. The difference to me is that it better represents bytes because most operations on a vec aren't useful when dealing with bytes. For example it may be useful to sort a Vec<u8>, it's probably not useful in most cases to sort a `Box<[u8]>`. It also let's you implement a trait for a Vec<T> and a Box<[u8]> as sometimes you wan't to handle bytes differently from a list. But I've mostly given up on this, as everyone just uses Vec<u8>.

Reply:
> as opposed to a Vec where the allocated capacity could be a lot bigger than the len.

yeah that makes sense.

Reply:
@dvc94ch [`into_boxed_slice`](https://doc.rust-lang.org/std/vec/struct.Vec.html#method.into_boxed_slice) drops the excess capacity, at least.

Reply:
if you want something that better represents bytes, why not just use the `bytes` crate?

Reply:
What's the advantage of using the bytes crate vs a Box<[u8]>?

Reply:
> What's the advantage of using the bytes crate vs a Box<[u8]>?

reference-counted so cheaper to clone.

Reply:
I don't have any strong opinions on this topic, I just have never seen a `Box<[u8]>` in the wild before and I thought it was a very clever way to abstract bytes.

Reply:
So what you want is an Arc<[u8]>. Anyway looks like into_boxed_slice does drop excess capacity, there used to be some issues with it https://github.com/rust-lang/rust/issues/36284, but I can't find why I thought it didn't.

Reply:
also, I just checked and the bytes crate has a lot of convenience methods for working with bytes :smile:  (LE/BE get/set).

Reply:
As I said, I might change it to just return a Vec<u8>, if people prefer. I've had a few people disagree on the utility of a Box<[u8]>. I probably won't force people to use a specific crate, you can easily convert a vec or a box to bytes.

Reply:
In this kind of situation, in my opinion, one should simply expose implementation details. If the implementation uses `Vec`, return a `Vec`, so that you're not copying bytes for no reason.